source/extensions/filters/http/gcp_authn/gcp_authn_filter.cc

```diff
@@ -0,0 +1,83 @@
+#include "source/extensions/filters/http/gcp_authn/gcp_authn_filter.h"
+
+#include "source/common/common/enum_to_int.h"
+#include "source/common/http/header_map_impl.h"
+#include "source/common/http/utility.h"
+
+#include "absl/strings/str_replace.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace GcpAuthn {
+
+using ::Envoy::Router::RouteConstSharedPtr;
+using Http::FilterHeadersStatus;
+
+Http::FilterHeadersStatus GcpAuthnFilter::decodeHeaders(Http::RequestHeaderMap&, bool) {
+  Envoy::Router::RouteConstSharedPtr route = decoder_callbacks_->route();
+  if (route == nullptr || route->routeEntry() == nullptr) {
+    // Nothing to do if no route, continue the filter chain iteration.
+    return Envoy::Http::FilterHeadersStatus::Continue;
+  }
+
+  state_ = State::Calling;
+  initiating_call_ = true;
+
+  Envoy::Upstream::ThreadLocalCluster* cluster =
+      context_.clusterManager().getThreadLocalCluster(route->routeEntry()->clusterName());
+
+  std::string audience_str;
+  if (cluster != nullptr) {
+    auto filter_metadata = cluster->info()->metadata().filter_metadata();
+    const auto filter_it = filter_metadata.find(std::string(FilterName));
+
+    if (filter_it != filter_metadata.end()) {
+      audience_str =
+          filter_it->second.fields().find(std::string(AudienceKey))->second.string_value();
+    }
+  }
+
+  if (!audience_str.empty()) {
+    // Add the audience from the config to the final url.
+    // `[AUDIENCE]` field is substituted with real audience string from the config.
+    std::string final_url =
+        absl::StrReplaceAll(filter_config_->http_uri().uri(), {{"[AUDIENCE]", audience_str}});
+    client_->fetchToken(*this, buildRequest("GET", final_url));
+    initiating_call_ = false;
+  } else {
+    // Audience is URL of receiving service that will performance authentication. So, there is no
+    // need to fetch the token if no audience is specified, because no authentication will be
+    // performed. Just continue the filter chain iteration.
+    ENVOY_LOG(error, "Failed to retrieve the audience from the configuration.");
```

Comment:
Actually I feel this log could be helpful : it helped me switch from metadata to typed metadata.

I agree it might be noisy so I downgrade the log severity from `error` to `debug`. Will this work?

Reply:
I have switched to stats. Done PTAL. Thanks!

The presubmit failure is caused by flaky filter_fuzzer_test that I cannot repro locally